File: backend/app/models/committee.py
"""Committee/venue ORM models."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Iterable

from ..extensions import db
from .user import User

COMMITTEE_STATUSES = ('preparation', 'in_session', 'paused', 'closed')


class Committee(db.Model):
    __tablename__ = 'Committees'

    id = db.Column(db.Integer, primary_key=True)
    code = db.Column(db.String(10), unique=True, nullable=False)
    name = db.Column(db.String(255), nullable=False)
    venue = db.Column(db.String(255))
    description = db.Column(db.Text)
    status = db.Column(
        db.Enum(*COMMITTEE_STATUSES, name='committee_status'),
        nullable=False,
        default='preparation',
    )
    capacity = db.Column(db.Integer, default=40)
    agenda_order = db.Column(db.JSON)
    dais_json = db.Column(db.JSON)
    time_config = db.Column(db.JSON)
    created_by = db.Column(db.Integer, db.ForeignKey('Users.id'))
    created_at = db.Column(db.DateTime, nullable=False,
                           default=datetime.utcnow)
    updated_at = db.Column(
        db.DateTime,
        nullable=False,
        default=datetime.utcnow,
        onupdate=datetime.utcnow,
    )

    sessions = db.relationship(
        'CommitteeSession',
        backref='committee',
        lazy='selectin',
        cascade='all, delete-orphan',
        order_by='CommitteeSession.start_time',
    )

    def to_dict(self) -> dict[str, Any]:
        return {
            'id': self.id,
            'code': self.code,
            'name': self.name,
            'venue': self.venue,
            'status': self.status,
            'capacity': self.capacity or 0,
            'description': self.description,
            'dais': self.dais,
            'sessions': [session.to_dict() for session in self.sessions],
            'timeConfig': self.time_config_data,
            'createdAt': self.created_at.isoformat() if self.created_at else None,
            'updatedAt': self.updated_at.isoformat() if self.updated_at else None,
        }

    @property
    def dais(self) -> list[dict[str, Any]]:
        if not self.dais_json:
            return []
        ids = [d['id']
               for d in self.dais_json if isinstance(d, dict) and 'id' in d]
        if not ids:
            return []
        users = {u.id: u for u in User.query.filter(User.id.in_(ids)).all()}
        return [
            {
                'id': d['id'],
                'name': users[d['id']].name if d['id'] in users else 'Unknown',
                'role': d.get('role', '主席团'),
                'contact': users[d['id']].phone if d['id'] in users else None,
            }
            for d in self.dais_json
            if isinstance(d, dict) and 'id' in d
        ]

    @dais.setter
    def dais(self, value: list[dict[str, Any]] | None) -> None:
        if value is None:
            self.dais_json = []
        else:
            self.dais_json = [{'id': int(v['id']), 'role': str(
                v.get('role', '主席团'))} for v in value]

    @property
    def time_config_data(self) -> dict[str, Any]:
        if isinstance(self.time_config, dict):
            return self.time_config
        return {'realTimeAnchor': None, 'flowSpeed': 1}

    @time_config_data.setter
    def time_config_data(self, value: dict[str, Any]) -> None:
        self.time_config = value or None


class CommitteeSession(db.Model):
    __tablename__ = 'CommitteeSessions'

    id = db.Column(db.Integer, primary_key=True)
    committee_id = db.Column(
        db.Integer, db.ForeignKey('Committees.id', ondelete='CASCADE'), nullable=False, index=True)
    topic = db.Column(db.String(255), nullable=False)
    chair = db.Column(db.String(255))
    start_time = db.Column(db.DateTime)
    duration_minutes = db.Column(db.Integer, nullable=False, default=20)
    created_at = db.Column(db.DateTime, nullable=False,
                           default=datetime.utcnow)
    updated_at = db.Column(
        db.DateTime,
        nullable=False,
        default=datetime.utcnow,
        onupdate=datetime.utcnow,
    )

    def to_dict(self) -> dict[str, Any]:
        return {
            'id': str(self.id),
            'topic': self.topic,
            'chair': self.chair,
            'start': self.start_time.isoformat() if self.start_time else None,
            'durationMinutes': self.duration_minutes,
        }


__all__ = ['Committee', 'CommitteeSession', 'COMMITTEE_STATUSES']

File: backend-php/app/Models/Committee.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Collection;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Support\Carbon;

class Committee extends Model
{
    protected $table = 'Committees';

    protected $fillable = [
        'code',
        'name',
        'venue',
        'description',
        'status',
        'capacity',
        'agenda_order',
        'dais_json',
        'time_config',
        'created_by',
    ];

    protected $casts = [
        'created_at' => 'datetime',
        'updated_at' => 'datetime',
    ];

    public function sessions(): HasMany
    {
        return $this->hasMany(CommitteeSession::class, 'committee_id')->orderBy('start_time');
    }

    public function creator(): BelongsTo
    {
        return $this->belongsTo(User::class, 'created_by');
    }

    /**
     * @return array<int, array<string, mixed>>
     */
    public function daisMembers(): array
    {
        $raw = $this->dais_json;
        if (!$raw) {
            return [];
        }
        $decoded = is_array($raw) ? $raw : json_decode((string) $raw, true);
        if (!is_array($decoded)) {
            return [];
        }
        $ids = [];
        foreach ($decoded as $item) {
            if (is_array($item) && isset($item['id'])) {
                $ids[] = (int) $item['id'];
            }
        }
        $users = [];
        if ($ids) {
            $users = User::query()->whereIn('id', $ids)->get()->keyBy('id')->all();
        }
        $result = [];
        foreach ($decoded as $item) {
            if (!is_array($item) || !isset($item['id'])) {
                continue;
            }
            $id = (int) $item['id'];
            $user = $users[$id] ?? null;
            $result[] = [
                'id' => $id,
                'name' => $user?->name ?? 'Unknown',
                'role' => $item['role'] ?? '主席团',
                'contact' => $user?->phone,
            ];
        }
        return $result;
    }

    /**
     * @param array<int, array{id:int|string, role?:string}> $value
     */
    public function setDaisMembers(array $value): void
    {
        $normalized = array_map(static function ($item) {
            return [
                'id' => (int) ($item['id'] ?? 0),
                'role' => (string) ($item['role'] ?? '主席团'),
            ];
        }, $value);
        $this->dais_json = json_encode($normalized, JSON_UNESCAPED_UNICODE);
    }

    /**
     * @return array<string, mixed>
     */
    public function timeConfig(): array
    {
        $raw = $this->time_config;
        if (is_array($raw)) {
            return $raw;
        }
        $decoded = json_decode((string) $raw, true);
        if (is_array($decoded)) {
            return $decoded;
        }
        return ['realTimeAnchor' => null, 'flowSpeed' => 1];
    }

    public function setTimeConfig(?array $config): void
    {
        $this->time_config = $config ? json_encode($config, JSON_UNESCAPED_UNICODE) : null;
    }

    /**
     * @param Collection<int, CommitteeSession>|null $sessions
     * @return array<string, mixed>
     */
    public function toApiResponse(?Collection $sessions = null): array
    {
        $sessions = $sessions ?? $this->sessions;
        $formatDate = static fn ($value) => $value ? Carbon::parse($value)->toIso8601String() : null;

        return [
            'id' => $this->id,
            'code' => $this->code,
            'name' => $this->name,
            'venue' => $this->venue,
            'status' => $this->status,
            'capacity' => (int) ($this->capacity ?? 0),
            'description' => $this->description,
            'dais' => $this->daisMembers(),
            'sessions' => $sessions->map(static fn (CommitteeSession $session) => $session->toApiResponse())->all(),
            'timeConfig' => $this->timeConfig(),
            'createdAt' => $formatDate($this->created_at),
            'updatedAt' => $formatDate($this->updated_at),
        ];
    }
}
